tls/ttls.c

```diff
@@ -149,7 +151,6 @@ ttls_crypto_req_sglist(TlsCtx *tls, struct crypto_aead *tfm, unsigned int len,
 		off = io->off;
 		n = *sgn + io->chunks;
 	}
-	BUG_ON(!buf && (!skb || skb->len <= off)); /* nothing to decrypt */
```

Comment:
> When do we have skb->len == 0?

This BUG_ON was triggering on `skb->len` of 10 bytes while `off` was 13. I don't think the issue was with zero-length skbs. This is the place before skb chain trimming, so it's expected to have `skb->len` less than `off`. The code below handles that case explicitly, so I think the condition was more strict than it needed to be. We still need to ensure that either `buf` or `skb` is non-NULL.

> Probably, this is job for #1330

I'm inclined to keep trimming-related fixes outside this PR, and believe that #1330 (or a separate PR) is a better place for that changes.

Reply:
There is no need to handle `!buf && !skb` case, no null derefence will happen. The `skb` is used inside the `for` loop, where the condition `skb != NULL` is evaluated before entering the loop. So it's OK. Condition `skb->len <= off` is really buggy here, so I;m generally OK with the change.

But one thing is bothering me. If `off` is outside the skb list (e.g. total size of skbs is 10, while the offset is 13), or if `!buf && !skb = true`, the function will return the `req` to the caller - `__ttls_decrypt()`. In this case it's very possible that `sgn` will be set to `1`. Condition `if (!req || WARN_ON_ONCE(sgn < 2))` in `__ttls_decrypt()` will fail and a error will be generated. But the `req` is not destroyed and memory leak happens. We need to properly release `req` on errors. The leakage is not possible on master since the `BUG_ON` statement is more strict that it's required.

Reply:
> 	WARN_ON_ONCE(len == 0); /* nothing to decrypt */

Seems it never happen. The `ttls_crypto_req_sglist()` is called only from `__ttls_decrypt()`. According to the later: https://github.com/tempesta-tech/tempesta/blob/0483d05ac48d6d55090913ddf874181a13dd9c19/tls/ttls.c#L880-L892

So we get `len = dec_msglen + taglen = io->msglen - expiv_len - taglen + taglen =  io->msglen - expiv_len`. But the condition above guarantee that the `io->msglen >= expiv_len + taglen`, and `taglen` is always positive non-null value. Thus `len >= taglen` and always positive, and `WARN_ON_ONCE(len == 0)`  never happens.